version: "3.7"

configs:
  nginx.conf:
    name: "nextcloud_nginx-${DEPLOY_TIMESTAMP}.conf"
    file: "./nginx.conf"

services:
  nextcloud: &nextcloud
    image: "nextcloud:{{ nextcloud_version }}-fpm-alpine"
    volumes:
      - nextcloud-data:/var/www/html
    environment:
      # Database
      - "MYSQL_DATABASE=${MYSQL_DATABASE}"
      - "MYSQL_USER=${MYSQL_USER}"
      - "MYSQL_PASSWORD=${MYSQL_PASSWORD}"
      - "MYSQL_HOST=${MYSQL_HOST}"
      # Nextcloud
      - "NEXTCLOUD_ADMIN_USER=${NEXTCLOUD_ADMIN_USER}"
      - "NEXTCLOUD_ADMIN_PASSWORD=${NEXTCLOUD_ADMIN_PASSWORD}"
      # Mailing
      - "SMTP_HOST=${SMTP_HOST}"
      - "SMTP_SECURE=${SMTP_SECURE}"
      - "SMTP_PORT=${SMTP_PORT}"
      - "SMTP_AUTHTYPE=${SMTP_AUTHTYPE}"
      - "SMTP_NAME=${SMTP_NAME}"
      - "SMTP_PASSWORD=${SMTP_PASSWORD}"
      - "MAIL_FROM_ADDRESS=${MAIL_FROM_ADDRESS}"
      - "MAIL_DOMAIN=${MAIL_DOMAIN}"
    networks:
      - nextcloud-network
    deploy:
      placement:
        constraints:
          - node.role == manager
          - node.labels.nextcloud.nextcloud-data == true
      restart_policy:
        condition: on-failure
        delay: 5s
        window: 60s
        max_attemps: 3

  nextcloud-cron-worker:
    <<: *nextcloud
    deploy:
      mode: replicated
      replicas: 0
      restart_policy:
        condition: none
      labels:
        - "swarm.cronjob.enable=true"
        - "swarm.cronjob.schedule=0 */5 * * * *"
        - "swarm.cronjob.skip-running=true" # do not start a job is one already exists
    command: su -l www-data -s /bin/sh -c 'php -f /var/www/html/cron.php'

  nginx:
    image: nginx
    volumes:
      - nextcloud-data:/var/www/html:ro
    networks:
      - nextcloud-network
      - {{ traefik_network }}
    configs:
      - source: nginx.conf
        target: /etc/nginx/nginx.conf
    deploy:
      placement:
        constraints:
          - node.role == manager
          - node.labels.nextcloud.nextcloud-data == true
      restart_policy:
        condition: on-failure
        delay: 5s
        window: 30s
        max_attemps: 3
      labels:
        - "traefik.frontend.rule=Host:{{ nextcloud_domain }}"
        - traefik.enable=true
        - traefik.port=80
        - "traefik.tags={{ traefik_public_tag }}"
        - "traefik.docker.network={{ traefik_network }}"
        # Traefik service that listens to HTTP
        - traefik.redirectorservice.frontend.entryPoints=http
        - traefik.redirectorservice.frontend.redirect.entryPoint=https
        # Traefik service that listens to HTTPS
        - traefik.webservice.frontend.entryPoints=https
        # HSTS
        - traefik.frontend.headers.SSLRedirect=true
        - "traefik.frontend.headers.SSLHost={{ nextcloud_domain }}"
        - traefik.frontend.headers.STSSeconds=315360000
        - traefik.frontend.headers.STSIncludeSubdomains=true
        - traefik.frontend.headers.STSPreload=true
        # security labels
        - traefik.frontend.headers.browserXSSFilter=true
        - traefik.frontend.headers.customFrameOptionsValue=sameorigin # allow same origin for iframes
        - traefik.frontend.headers.contentTypeNosniff=true
        # rate limits
        - traefik.frontend.rateLimit.extractorFunc=client.ip
        - traefik.frontend.rateLimit.rateSet.rate3s.period=3s
        - "traefik.frontend.rateLimit.rateSet.rate3s.average={{ nextcloud_max_requests_3s_per_client|int }}"
        - "traefik.frontend.rateLimit.rateSet.rate3s.burst={{ nextcloud_max_requests_3s_per_client * 2|int }}"
        # max connections limits
        - "traefik.backend.maxconn.amount={{ nextcloud_max_connections|int }}"
        - traefik.backend.maxconn.extractorFunc=request.host
        # requests size limits: currently disabled due to a bug with 302 requests
        # - "traefik.backend.buffering.maxRequestBodyBytes={{ nextcloud_max_request_size_megabytes * 1000000|int }}" # maximum request size allowed
        # - traefik.backend.buffering.memRequestBodyBytes=2097152 # after this limit, request will be buffered on disk instead of RAM
        # - traefik.backend.buffering.retryExpression=IsNetworkError() && Attempts() <= 2
        # Nextcloud CalDAV
        - traefik.frontend.redirect.permanent=true
        - traefik.frontend.redirect.regex=https://(.*)/.well-known/(card|cal)dav
        - traefik.frontend.redirect.replacement=https://$$1/remote.php/dav/

networks:
  nextcloud-network:
    driver: overlay
    driver_opts:
      encrypted: ""
  {{ traefik_network }}:
    external: true

volumes:
  nextcloud-data:
    driver: local
